src/prods/p10.py

```diff
@@ -0,0 +1,155 @@
+        node_h = graph.node_for_handle(rev_mapping[6])
+
+        # check vertices' labels
+        for node in vertices:
+            if node.attrs.label != 'v' or node.attrs.hanging:
+                return False
+
+        # in case of inner node, the `hanging' parameter means that the
+        # polygon is chosen to be broken; it is the same as the R parameter 
+        if not node_p.attrs.hanging or node_p.attrs.label != 'p':
+            return False
+
+        if not node_h.attrs.hanging:
+            return False
+
+        # verify edges coming out of the inner node
+        for node in vertices:
+            edge = graph.edge_for_handles(node.handle, node_p.handle)
+            if edge.attrs.kind != 'p':
+                return False
+
+        idx_nodes = [1, 4, 2, 3, 0]
+        nodes_shuffled = [vertices[idx] for idx in idx_nodes]
+        for node_a, node_b in it.pairwise([node_h] + nodes_shuffled + [node_h]):
+            edge = graph.edge_for_handles(node_a.handle, node_b.handle)
+            if edge.attrs.kind != 'e':
+                return False
+
+        return True
+
+    def apply_with_mapping(self, graph: Graph, mapping: Dict[NodeHandle, NodeHandle]):
+        # Break the edges (1, 2), (2, 3), (3, 4), (4, 1), creating 4 new nodes
+        # Add 5 new node to the center of the split
+        self.rev_mapping = util.reverse_dict_mapping(mapping)
+
+        # Aliasing for convenience
+        rev_mapping = self.rev_mapping
+
+        # get square's vertices
+        vertices = [
+            graph.node_for_handle(rev_mapping[i]) for i in range(0, 5)
+        ]
+        # get inner node, the hyperedge
+        node_p = graph.node_for_handle(rev_mapping[5])
+        node_h = graph.node_for_handle(rev_mapping[6])
+
+        # you must get edges attrs before you remove, now it is relevant
+        attrs = graph.edge_attrs((vertices[0].handle, node_h.handle))
+        is_boundry = attrs.value
+        x = node_h.attrs.x
+        y = node_h.attrs.y
+        hdl = node_h.handle
+        graph.remove_node(node_h.handle)
+        # change hanging atribute of the node
+        hang = Node(NodeAttrs('v', x, y, hanging=False), hdl)
+        graph.add_node(hang)
+        # recreate edges with proper boundry property
+        new_attr = EdgeAttrs('e', is_boundry)
+        edge = Edge(vertices[0].handle, hdl, new_attr)
+        graph.add_edge(edge)
+        new_attr = EdgeAttrs('e', is_boundry)
+        edge = Edge(vertices[1].handle, hdl, new_attr)
+        graph.add_edge(edge)
+
+        # skip first pair because it already exists
+        idx_nodes = [1, 4, 2, 3, 0]
+        nodes_shuffled = [vertices[idx] for idx in idx_nodes]
+
+        # hanging node is already created, should be on boundry list
+        new_boundry = [hang]
+        for node_a, node_b in it.pairwise(nodes_shuffled):
+            x = (node_a.attrs.x + node_b.attrs.x) / 2
+            y = (node_a.attrs.y + node_b.attrs.y) / 2
```

Comment:
Everywhere you compute coordinates you can take advantage of [`util.avg_points_from_nodes`](https://github.com/kkafar/graph-grammars/blob/759ae07b50a04c02d630ab37fec84ff3b29b45bb/src/util.py#L32) (I believe it was also added after you opened the PR)